### DISCUSSION TITLE
How to build the file system of monaco editor?

### DISCUSSION THREAD

**Discussion by newfish-cmyk** [TypeFox/monaco-languageclient#551] [Q&A] [answered]
I created 2 model, the `index2.ts` with code below,
![image](https://github.com/TypeFox/monaco-languageclient/assets/71265218/899b810e-b687-4f20-a789-f797a3ec5fa9)
and in `index.ts`, I can auto import like picture below
![image](https://github.com/TypeFox/monaco-languageclient/assets/71265218/667e3c3f-6726-4b13-a18c-2b611358242a)
but with error 
![image](https://github.com/TypeFox/monaco-languageclient/assets/71265218/63c1d55c-9d8c-4f8b-8601-909a15809e3e)

when I use `import "./"`, it shows options below, which are the local files
![image](https://github.com/TypeFox/monaco-languageclient/assets/71265218/8cba4a67-ec92-4dc7-ba5c-005bb6622b49)

So my problem is how can I change the "./" options from local files to virtual model files? 
Thank you for all the help! 🌹

[Demo](https://github.com/newfish-cmyk/monaco-lsp-demo)

**Comment by CGNonofr** [comment, marked as answer]
The language server runs on the real filesystem, so you either need a way to access the server filesystem from the client, or synchronize the client changes on the filesystem

There is multiple way of doing it:
- just watching the model change and writing them to the disk (not recommended)
- implement and registering a filesystem overlay (using `registerFileSystemOverlay` from `codingame/monaco-vscode-files-service-override`) (nice but not trivial to implement, requires a channel and a protocol)
- use the vscode ext-host server (fully automatic, but has some downside: it requires its own websocket, it needs to be up and running from the beginning, you can't disconnect/reconnect to it...)

**Reply by newfish-cmyk** [reply]
Thank you for your timely response!
I'm trying to use the second method to create the filesystem overlay. 
Through `console.log("fileSystemProvider", fileSystemProvider.readdir(vscode.Uri.file("C:/Users/heheer/github/monaco-lsp/")));`
![image](https://github.com/TypeFox/monaco-languageclient/assets/71265218/b19e5e56-eb4e-4b7c-8c77-1981402dfae5)
I feel like the files have been created, but in fact it not working. 
May I ask if I'm missing the synchronize step? and how can I fix it?

**Reply by CGNonofr** [reply]
Where does the fileSystemProvider come from? you're probably only reading from the virtual filesystem

**Reply by newfish-cmyk** [reply]
I defined fileSystemProvider through useState.
```
  const [fileSystemProvider] = useState<RegisteredFileSystemProvider>(() => {
    const provider =  new RegisteredFileSystemProvider(false)
    registerFileSystemOverlay(1, provider);
    return provider
  });
```
and regist them by 

```
  allFunctionList.forEach(async (item: any) => {
    const path = `C:/Users/heheer/github/monaco-lsp/${item.name}`
    if (!monaco.editor.getModel(monaco.Uri.file(path))) {
      fileSystemProvider.registerFile(new RegisteredMemoryFile(vscode.Uri.file(path), item.source.code));
      const model = monaco.editor.createModel(item.source.code, "typescript", monaco.Uri.file(path))
      editorRef.current?.setModel(model);
    }
  })
```
It seems that I only read the virtual file system. 

So in fact the language server and monaco model belong to different file systems. What I need to do is to synchronize and overlay the virtual filesystem of the monaco model onto the system of the language server?

**Reply by CGNonofr** [reply]
Do not use RegisteredFileSystemProvider but implement your own filesystem provider (interface IFileSystemProviderWithFileReadWriteCapability) that reads and writes from/to the disk using any mean you want

**Reply by CGNonofr** [reply]
depending on your use-case, you can maybe use the `HTMLFileSystemProvider` as well. it uses the javascript filesystem api to access the filesystem of the machine where the browser runs (it ask for user permissions though, and not supported on Firefox and Safari)

**Reply by newfish-cmyk** [reply]
Thanks so much for your kind help! 
I finally managed to get it work properly (using `registerFileSystemOverlay` from `codingame/monaco-vscode-files-service-override`)

![1697779565375](https://github.com/TypeFox/monaco-languageclient/assets/71265218/5f4044a2-5b05-480a-bebc-cb9a2c63fbcc)

**Reply by kaisalmen** [reply]
@newfish-cmyk superb! 🎉 Are you interested to add this as another example to this repo and create a PR for it? It could be a nice addition to the python LS example. @CGNonofr WDYT?

**Reply by CGNonofr** [reply]
Totally agree!

**Comment by vivekburman** [comment]
Hi @newfish-cmyk can you share your solution repo?

**Comment by VamshiGoud** [comment]
@newfish-cmyk  Please share the solution, working on the same thing.
